Proposal to add `oci-artifact` as pURL type (#123)
package-url/purl-spec: PURL-TYPES.rst

      pkg:deb/debian/dpkg@1.19.0.4?arch=amd64&distro=stretch
      pkg:deb/ubuntu/dpkg@1.19.0.4?arch=amd64

> docker

Reviewer: The removal of 'docker' as a valid purl type will break compatibility for millions of users who have been using this type for the past several years.

Author: Ok, we can leave docker in to be backwards compatible. Would you be OK with adding a comment within the docker section suggesting to use `oci-artifact` instead of `docker`? A docker image is a type of oci artifact.

Reviewer 2: If I may: the addition of the `docker` portion of the spec was added in 2021 by @iamwillbar (and I would hazard a guess that it was after previous conversations on how to represent container images in the purl spec). `oci-artifact` is the packaging format used by containers built by tools like docker. This change has been discussed with the OCI community here: https://hackmd.io/El8Dd2xrTlCaCG59ns5cwg#September-8-2021

Reviewer 3: @nishakm that's not quite correct, it was added in 2017: https://github.com/package-url/purl-spec/commit/cdf28ffb1cd91424d823dcd591999fbca0f10d8f. My change was purely to the formatting of the document.

Reviewer 2: The commit adding the "docker" examples:
```
commit 2197344e22a93d715256228b4a0e2069507410d7
Author: William Bartholomew <iamwillbar@github.com>
Date:   Tue Jun 15 13:51:43 2021 -0700

    Formatting of PURL-TYPES.rst
    
    Formatting changes, no semantic changes intended:
    * Adds headings for each type so that hyperlinks are automatically generated. Resolves #106.
    * Replaces backticks with double backticks as required by reStructuredText.
    * Adds the occasional Oxford comma for good measure.

```

Reviewer 3: @nishakm those are from the 2017 version of the file: https://github.com/package-url/purl-spec/commit/cdf28ffb1cd91424d823dcd591999fbca0f10d8f#diff-7b3ed02bc73dc06b7db906cf97aa91dec2b2eb21f2d92bc5caa761df5bbc168fR460. I just moved them and re-formatted them 🙂 .

Reviewer 2: Thanks @iamwillbar! I stand corrected about the date and original author. @pombredanne, Do you remember where you got the notion of adding "docker" as the package-type?

Reviewer 2: PS: I am OK leaving "docker" in the spec for now, provided it is understood that the canonical repository is "hub.docker.com".

Reviewer 3: The spec does define the default repository as `hub.docker.com`, but it does allow a different registry to be specified with the `repository_url` parameter. It might be worth having a "See Also" in the `docker` section that refers to the new `oci-artifact` as a more general option.

Reviewer 2: So this is the rub for including `docker` as a `package-type` at all. Even the tool "Docker" uses the OCI image specification as the format. There is no canonical container image repository. This ecosystem might as well be synonymous with storing `.exe` files on Sourceforge. If I can extend that analogy, the `oci-artifact` is like `http` (the transport) and `docker` is like `wget` (the command line tool).

Reviewer 4: when I read "docker" for docker compat registries I read docker v1 registry api and docker v2 registry api (near 1/1 with oci distribution api) for docker schema format container images and also depending on release of the docker cli and hub/registry code probably also supporting oci schema formatted objects 

When I read oci for oci compatible registries images, I read more focused on just the oci v1 distribution api for storing all of the oci v1 base types (index/manifest/blob/header/config...) and not to include the similar docker v2 format.

It should be noted that docker v1 isn't widely used anymore with the focus on container images now being on docker v2 and oci v1.

      pkg:nuget/EnterpriseLibrary.Common@6.0.1304

> oci-artifact

Reviewer 5: If container images are in scope for this, then I would consider renaming this purl type to just `oci`. Not all images conform to [OCI Artifacts](https://github.com/opencontainers/artifacts); e.g. multi-platform images don't fit into that model currently.

Reviewer 6: See [above](https://github.com/package-url/purl-spec/pull/123#discussion_r710606060) where we agreed this would be oci-artifact as images are one of many types of artifacts.

Reviewer 4: naming is as hard as scoping :-) I agree with @jonjohnsonjr if the scope was docker vs oci images we should name it oci or oci-image.

I think longer term though we should probably just scope this as all oci-artifact types stored in one of these oci compatible registries.  I think that's best for the OCI consumer?  borrowing the definitions from the distribution spec:
```
- **Registry**: a service that handles the required APIs defined in this specification
- **Client**: a tool that communicates with Registries
- **Push**: the act of uploading Blobs and Manifests to a Registry
- **Pull**: the act of downloading Blobs and Manifests from a Registry
- **Blob**: the binary form of content that is stored by a Registry, addressable by a Digest
- **Manifest**: a JSON document which defines an Artifact. Manifests are defined under the OCI Image Spec <sup>[apdx-2](#appendix)</sup>
- **Config**: a blob referenced in the Manifest which contains Artifact metadata. Config is defined under the OCI Image Spec <sup>[apdx-4](#appendix)</sup>
- **Artifact**: one conceptual piece of content stored as Blobs with an accompanying Manifest containing a Config
- **Digest**: a unique identifier created from a cryptographic hash of a Blob's content. Digests are defined under the OCI Image Spec <sup>[apdx-3](#appendix)</sup>
- **Tag**: a custom, human-readable Manifest identifier
```
in that context(scope) .. oci-artifacts would mean conceptual pieces of content stored as OCI Blobs with an accompanying OCI Manifest containing a OCI Config

------------

> ``oci-artifact`` for all artifacts stored in OCI Distribution-spec conformant

Reviewer 3: Can we link to the OCI Distribution-spec here?

Reviewer 7: How about this, @iamwillbar?

```suggestion
``oci-artifact`` for all artifacts stored in registries that conform to the `OCI
Distribution Specification <https://github.com/opencontainers/distribution-spec>`_,
including container images built by Docker and others:
```

> registries including container images built by docker and others:

Reviewer 3: ```suggestion
registries including container images built by Docker and others:
```

Reviewer 6: Just for context, Docker is used to refer to **Docker Inc**, while docker is used to refer to the `docker` cli. When referring to process containers, we've tried to shift to _container images_ to be vendor neutral.

- There is no canonical package repository for OCI artifacts. Therefore
  ``oci-artifact`` pURLs must be registry agnostic.
- OCI pURLs do not require a ``namespace``. The ``namespace`` is specific
  to the physical location of the package.
- The ``name`` is not case sensitive and must be lowercased. The name is the
  last fragment of the repository name. For example if the repository
  name is ``library/debian`` then the ``name`` is ``debian``.
- The ``version`` is the ``@sha256:digest`` of the artifact and is strongly

>   suggested for oci artifacts.

Reviewer 3: ```suggestion
  suggested for OCI artifacts.
```

- Optional qualifiers may include:

  - ``arch``: key for a package architecture, when relevant

>   - ``repository``: A repository URL where the artifact may be found, but not intended as the only location

Reviewer 3: Could we use `repository_url` to be consistent with other types?

Author: Works for me

Reviewer 8: Is this a URL or a repository name, and if the latter, does it include the registry? A repository name would be `example.com:5000/group/app`, where the URL would need the schema in front (e.g. `https://`) which isn't normally seen in container image references.

Author: If you take a look in the examples, we are proposing something like `docker.io/library/debian` for this value.

Reviewer 8: So instead of "repository URL", I'd recommend calling it the "OCI repository". Calling it a URL implies to me I could `curl docker.io/library/debian`, which isn't a feature of OCI.

  - ``tag``: artifact tag that may have been associated with the digest at the time

>   - ``type``: type of artifact stored in an OCI registry

Reviewer 3: Is this always a Media Type?

Author: Yes, will always be a media type. I can clarify in the description.

Reviewer 6: It's a little more nuanced. We [initially used media types](https://github.com/opencontainers/artifacts/blob/master/artifact-authors.md#defining-a-unique-artifact-type) to squeeze into the existing `manifest.config.mediatype` format.
Moving forward, we use [artifactType](https://github.com/oras-project/artifacts-spec/blob/main/artifact-manifest.md#oras-artifact-manifest-properties) and have decoupled from `manifest.config.mediaType`. 
However, to avoid artifact authors from having to maintain two types, we are using iana.org mediaTypes to uniquely differentiate types. 
> For details on creating a unique artifactType, see [OCI Artifact Authors Guidance](https://github.com/opencontainers/artifacts/blob/master/artifact-authors.md)

Reviewer 5: Using `type` here is confusing, given that purl already defines a `type`.

This is also ambiguous.

Does `type` refer to the media type of the artifact, the config mediaType (as [defined in OCI artifacts](https://github.com/opencontainers/artifacts/blob/master/artifact-authors.md#defining-oci-artifact-types)), or the `artifactType` from [oras artifacts](https://github.com/oras-project/artifacts-spec/blob/main/artifact-manifest.md#oras-artifact-manifest-properties)?

If the first, I'd suggest renaming to `mediaType`.

If the second, `artifactType` seems fine.

If the third, I think the purl `type` should probably be renamed to `oras-artifact`, given that they don't have a compatible schema with OCI Artifacts.

Author: > Using `type` here is confusing, given that purl already defines a `type`.

type is used by other purl types (see: maven) so I don't think this is a conflict with the purl spec but I don't have issues renaming it to `mediaType`.

Reviewer 5: > type is used by other purl types (see: maven)

Oh cool, that's fine with me then. Can you clarify if this refers to the artifact's [media type](https://github.com/opencontainers/image-spec/blob/main/media-types.md) or its [artifact type](https://github.com/opencontainers/artifacts/blob/master/artifact-authors.md#defining-oci-artifact-types)?

Author: It should refer to the artifact's artifact type as defined [here](https://github.com/opencontainers/artifacts/blob/master/artifact-authors.md#defining-oci-artifact-types). This is illustrated in some of the examples above as well.

Reviewer 5: In that case, I don't think this PR is about OCI artifacts in general, but specifically about OCI Artifacts, which wouldn't completely encompass container images.

Reviewer 6: > Oh cool, that's fine with me then. Can you clarify if this refers to the artifact's media type or its artifact type?

@jonjohnsonjr, this was also your feedback to use the same format for both `manifest.config.mediaType` and `oras.manifest.artifactType`

SInce mediaType has such broad meanings, we were using artifactType, which happens to use the iana mediaType format.

Reviewer 8: If this is the `artifactType` field from an OCI Artifact, then are we saying this purl does not apply to OCI Images (and docker images) since those do not have an `artifactType` field? Or is there a decision tree based on some logic that maps the current manifest media type + config media type to this `type` field?

Reviewer 5: > If this is the `artifactType` field from an OCI Artifact
 
The `artifactType` field is from ORAS Artifacts, not OCI Artifacts.

> - ``subpath`` is not relevant to ``oci-artifact``

Reviewer 3: I don't think you need to include optional qualifiers that aren't relevant, I'd remove this line.

- Examples::

      pkg:oci-artifact/debian@256:<shasum>?repository=docker.io/library/debian&arch=amd64&type=application/vnd.oci.image.manifest.v1+json
      pkg:oci-artifact/debian@sha256:<shasum>?tag=bullseye&type=application/vnd.oci.image.manifest.v1+json
      pkg:oci-artifact/metallb@sha256:<shasum>?repository=registry.io/bitnami&type=application/vnd.cncf.helm.config.v1+json

>       pkg:oci-artifact/hello-wasm@256:<shasum>?type=application/vnd.wasm.config.v1+json&tag=v1

Reviewer: Per the PURL spec, PURL is used to "reliably identify and locate software"

I don't see how anyone can locate using this purl.

Author: Is the `repository` URL identifier not enough?

Reviewer 2: This is a valid concern, and also something I've discussed with @iamwillbar several times: In cloud native, there isn't a single ecosystem akin to other package managers. For example, there is no "docker" ecosystem similar to a "pip" ecosystem. Container registries are distributed, there are many client tools, and the artifacts can be any number of things. The only thing that ties them together is the OCI specification. Hence the "oci-artifact" type.
We would also like to not link the pURL to a specific domain or tool as they can move from cloud to cloud and still be the same artifact.

Reviewer: I understand the distributed nature of oci artifacts. My point is with line 236, which does not have a `repository` specified and oci-artifact does not have a default. Therefore it is impossible to locate the example specified on line 236

```pkg:oci-artifact/hello-wasm@256:<shasum>?type=application/vnd.wasm.config.v1+json&tag=v1```

IMO, repository should be mandatory for oci-artifact purl types. Without it, they're not formatted in the spirit of purl which is to "reliably identify and locate software"

Reviewer 3: @stevespringett do you think it could be "strongly recommended" that one location to retrieve the artifact is provided in the `repository_url`. There could be scenarios where you want to share the identity but not a location (e.g. the location is currently an internal/private location or not yet known, the advantage of that is that adding a `repository_url` later doesn't invalidate the identity portion of the purl.

Reviewer: If private repos are used, they should be specified. Without a repository_url (or a default) it will be impossible to locate the artifact, so organizations will not be able to perform any kind of evaluation of the artifact (outdated, tampered, verification, etc).  

This ticket was framed in the context of SBOM, and security is the only reason why we're discussing SBOM. Therefore, if a new purl type is supported without the ability to locate the artifact, we won't be able to support security use cases. This doesn't seem like a good strategic decision. 

To my understanding, the only reason why oci-artifact is necessary is to support artifacts other than container images. Everything else can easily be supported with the existing Docker type with minor changes.  IMO, we may want to position oci-artifact as a superset of Docker which means oci-artifact should inherit Dockers defaults, including repository_url. If a NULL repository_url is truly wanted, then I think we need to define an optional way to specify that which would be applicable to every purl type, but it should not be the default simply by omitting repository_url as is the current case with this PR.

Author: > the only reason why oci-artifact is necessary is to support artifacts other than container images

It's true that we need oci-artifacts to support artifacts other than container images but also, describing all container images as `docker` is not correct. Docker is not a container image format, it's just the build tool.  As @nishakm mentioned above, even Docker uses the OCI image specification as the format and we should properly refer to this in purl.

> If a NULL repository_url is truly wanted, then I think we need to define an optional way to specify that which would be applicable to every purl type

No objection on this -- let's make this possible! A NULL repository has use cases outside of just oci artifacts as private repositories exist for other package types as well.

Reviewer 3: I'm hesitant for other ecosystems where the identity isn't a content address like it is for `oci-artifact`. If I take a package identity to two different registries I could get different packages back for some ecosystems, so I'd recommend only allowing  `repository_url` to be excluded if the package manager's purl has some way of guaranteeing it's the same package.

A NULL repository for `oci-artifact` may make sense when building the SBOM before you've published to a registry and don't know the "final" destination for the image. As I mentioned above a `repository_url` can be added without breaking the package identity, so updated SBOMs could be issued downstream when a specific canonical registry is known.

Author: > I'm hesitant for other ecosystems where the identity isn't a content address like it is for `oci-artifact`. If I take a package identity to two different registries I could get different packages back for some ecosystems, so I'd recommend only allowing `repository_url` to be excluded if the package manager's purl has some way of guaranteeing it's the same package.

Fair enough. To me, the "strongly recommended" approach makes the most sense but want to address all concerns. Cloud native artifacts do not compare "apples to apples" to typical package types with package managers so a slight deviation from the "spirit of purl" for this feels reasonable. The proposal for oci-artifact still ensures that the software can be "reliably identified".

> A NULL repository for `oci-artifact` may make sense when building the SBOM before you've published to a registry and don't know the "final" destination for the image. 

Or when the repository is private and you're unable to include it for security reasons, in which case, it may remain absent. But, as you mentioned, this would not affect the package identity.

Author: OCI artifacts i.e. kinds of artifacts defined by OCI

Reviewer 5: In that case, I'd rename this type to `oci` to avoid confusion with OCI Artifacts.

Author: Would the oci type in this case only be used to describe only oci container images?

Reviewer 6: @jonjohnsonjr, this was discussed in the last oci meeting, and you agreed to use oci-artifact. OCI was too general. the oci container image type is logically one of the OCI artifact types.
